view/statistics-flow-certificates.js

```diff
@@ -0,0 +1,322 @@
+				}
+			});
+		});
+	});
+
+	return result;
+};
+
+var incrementDateByTimeUnit = function (date, mode) {
+	switch (mode) {
+	case 'weekly':
+		date.setUTCDate(date.getUTCDate() + 7);
+		break;
+	case 'monthly':
+		date.setUTCMonth(date.getUTCMonth() + 1);
+		break;
+	case 'yearly':
+		date.setUTCFullYear(date.getUTCFullYear() + 1);
+		date.setUTCMonth(0);
+		break;
+	default:
+		date.setUTCDate(date.getUTCDate() + 1);
+		break;
+	}
+
+	return date;
+};
+
+var calculateDurationByMode = function (dateFrom, dateTo, mode) {
+	var timeUnitsCount = 0, currentDate;
+	currentDate = new db.Date(dateFrom.getTime());
+	if (!dateTo) dateTo = new db.Date();
+
+	while (currentDate < dateTo) {
+		incrementDateByTimeUnit(currentDate, mode);
+		timeUnitsCount++;
+	}
+	return timeUnitsCount;
+};
+
+var decorateQueryHandlerConf = function (queryHandlerConf, pagination) {
+	var conf = queryHandlerConf.slice(0);
+
+	conf.push({
+		name: 'page',
+		ensure: function (value, resolvedQuery, query) {
+			var num, dateFrom, dateTo, mode, durationInTimeUnits, resolvedValue;
+			resolvedValue = value;
+			if (resolvedValue == null) resolvedValue = '1';
+			if (isNaN(resolvedValue)) throw new Error("Unrecognized page value " + JSON.stringify(value));
+			num = Number(resolvedValue);
+			if (!isNaturalNumber(num)) throw new Error("Unreconized page value " + JSON.stringify(value));
+			if (num < 1) throw new Error("Unexpected page value " + JSON.stringify(value));
+
+			dateFrom = resolvedQuery.dateFrom; //new db.Date(location.query.get('dateFrom').value);
+			dateTo   = resolvedQuery.dateTo || new db.Date();
+			mode     = resolvedQuery.mode;
+			durationInTimeUnits = calculateDurationByMode(dateFrom, dateTo, mode);
+			pagination.count.value =
+				resolvedQuery.pageCount = Math.ceil(durationInTimeUnits / itemsPerPage);
+			if (num > resolvedQuery.pageCount) throw new Error("Page value overflow");
+			pagination.current.value = num;
+			return num;
+		}
+	});
+
+	return conf;
+};
+
+exports['statistics-main'] = function () {
+	var queryHandler, data = new ObservableValue({}), pagination = new Pagination('/flow/');
+	queryHandler = setupQueryHandler(decorateQueryHandlerConf(queryHandlerConf, pagination),
+		location, '/flow/');
+	queryHandler.on('query', function (query) {
+		var serverQuery = copy(query), dateFrom, dateTo, mode
+		  , currentDate, offset, timeUnitsCount = 0, durationInTimeUnits, now, page;
+
+		now      = new db.Date();
+		dateFrom = query.dateFrom;
+		dateTo   = query.dateTo || new db.Date();
+		mode     = query.mode;
+		page     = query.page;
+
+		if (query.pageCount > 1) {
+			durationInTimeUnits = calculateDurationByMode(dateFrom, dateTo, mode);
+			offset = { from: ((page - 1) * itemsPerPage) };
+			offset.to = offset.from;
+			if ((durationInTimeUnits - offset.from) < itemsPerPage) {
+				offset.to += durationInTimeUnits - offset.from;
+			} else {
+				offset.to += itemsPerPage;
+			}
+
+			currentDate = new db.Date(dateFrom.getTime());
```

Comment:
Here same way, to have accurate results we should _floor_ currentDate initially to given unit.

Otherwise in e.g. `month` mode, if dateFrom is 20th of some month, we will always ask for period starting or ending at 20th of some month, even when we ask for page 3 of 5.